ideas/000-discoverable-trusted-server-nodes/README.md

```diff
@@ -0,0 +1,73 @@
+---
+id: 000-discoverable-trusted-server-nodes
+title: Discoverable And Trusted Server Nodes
+status: Draft
+created: 2018-06-07
+category: core
+lead-contributor: adambabik
+contributors:
+exit-criteria: yes
+success-metrics: yes
+clear-roles: no
+future-iterations: yes
+roles-needed:
+    - Smart Contract Developer
+    - Go Developer
+    - Clojure Developer
+    - UI/UX Designer
+okrs:
+    - "[PO]: Objective: Messaging is reliable"
+    - "[P2]: KR: Cluster can handle 500 concurrent users"
+---
+
+## Summary
+The idea here is to provide a way to make Mail Server nodes discoverable and trusted. This should work seamlessly for the end user.
+
+## Product Overview
+Currently, Mail Servers (and other planned server nodes like Push Notification servers) need to be hardcoded in the app code. It's possible to provide your own Mail Server in the app settings but that's not a scalable solution and it does not solve all issues.
+
+The outcome of this swarm are fully decentralized and discoverable Mail Servers. The Status app will have a mean to verify that a discovered Mail Server is trusted.
+
+In the next iterations, we may work on a way to introduce external providers to submit their own Mail Servers into the cluster and allow end users to accept these resources as trusted.
+
+## Product Description
+The work in this swarm can be divided into three areas:
+- (Infra) making Mail servers discoverable,
+- (Smart Contract) writing a smart contract to keep track of trusted Mail Servers,
+- (Client) mechanism of verifying whether a discovered Mail Server is trusted.
+
+As mentioned in the Product Overview, the smart contract should allow registering custom providers. The client will have an option to decide which providers it trust. Status will be a default provider and by default the app will trust it.
+
+Due to the fact that all Mail Servers are independent and keep the same history of messages, we will be able to remove all hardcoded Mail Server addresses from the client and rely only on discovered Mail Servers. This will allow Status to dynamically scale number of their the Mail Servers depending on the usage.
+
+**Note** Important thing is that this solution should be flexible for other server nodes. For example, consider Push Notification servers that should be discovered and trusted in the same way.
+
+## Requirements & Dependancies
+1. Before this swarm starts, https://github.com/status-im/status-go/issues/1006 should be resolved.
+
+## Minimum Viable Product
+
+Goal Date: TBD
+
+1. [ ] A smart contract that keeps track of providers and Mail Servers,
+1. [ ] Discoverable Mail Servers,
+1. [ ] Requests for messages are signed with a public key of a provider,
```

Comment:
what for?

Reply:
Why not? :) By default, we should encrypt the communication between nodes unless there is a good reason not to. Is there any reason this should not be encrypted?

Reply:
but you mentioned signature not encryption

Reply:
Ah right. It should be about encryption. Sorry, I will fix that.

Reply:
would be nice if we could implement a simple diffie-hellman to avoid exchanging symkeys by whisper

Reply:
Provider's public key can be stored in the contract so no need to exchange anything. 

However, I don't know how to sign a mail server so that it can be verified it belongs to a given provider.

Reply:
can the mail server just sign messages with its private key so that the client can verify the signature with the public key found in the smart contract?

Reply:
Sorry, I wrote `signed` while I should be `encrypted`. This point is about replacing symmetric key which is currently used to encrypt a request for messages. Instead of a symmetric key, an asymmetric key should be used.

Reply:
> Instead of a symmetric key, an asymmetric key should be used. It will allow a provider to public its public key and keep the private key secure. Clearly, using a symmetric key is not an option.

this sounds a bit confusing. encrypting payload with pub key is not a prerequisite for discoverable mail server. we have all available data to confirm that we are connected exactly with target mail server.

Reply:
It's not prerequisite but everything we send through the network should be encrypted by default. Are devp2p packets encrypted by default? What if messages will be intercepted? 

I know that nothing sensitive will be exchanged by if it's easy to encrypt that communication, I don't see why we shouldn't do it.

Reply:
@adambabik afaik devp2p packets are encrypted by default using sym-key AES-256 (key exchanged through handshake), so it would be encrypted and probably provide enough security. 
In general you could re-phrase the sentence so we can abstract this and leave the details for later, it seems like has little impact on the overall implementation.
`Requests for messages are securely encrypted to provide confidentiality`.
We can later decide and flesh out whether `devp2p` is ok or we need another layer on top of it.
What do you think?